crates/bevy_reflect/src/serde/de.rs

```diff
@@ -12,375 +20,539 @@ pub trait DeserializeValue {
+impl<'de> Deserialize<'de> for Ident {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
     where
-        E: de::Error,
+        D: serde::Deserializer<'de>,
     {
-        Ok(Box::new(v))
+        struct IdentVisitor;
+
+        impl<'de> Visitor<'de> for IdentVisitor {
+            type Value = Ident;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("identifier")
+            }
+
+            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+            where
+                E: Error,
+            {
+                Ok(Ident(value.to_string()))
+            }
+
+            fn visit_string<E>(self, value: String) -> Result<Self::Value, E>
+            where
+                E: Error,
+            {
+                Ok(Ident(value))
+            }
+        }
+
+        deserializer.deserialize_identifier(IdentVisitor)
     }
+}
 
-    fn visit_i64<E>(self, v: i64) -> Result<Self::Value, E>
-    where
-        E: de::Error,
-    {
-        Ok(Box::new(v))
+/// A general purpose deserializer for reflected types.
+///
+/// This will return a [`Box<dyn Reflect>`] containing the deserialized data.
+/// For non-value types, this `Box` will contain the dynamic equivalent. For example, a
+/// deserialized struct will return a [`DynamicStruct`] and a `Vec` will return a
+/// [`DynamicList`]. For value types, this `Box` will contain the actual value.
+/// For example, an `f32` will contain the actual `f32` type.
+///
+/// This means that converting to any concrete instance will require the use of
+/// [`FromReflect`], or downcasting for value types.
+///
+/// Because the type isn't known ahead of time, the serialized data must take the form of
+/// a map containing the following entries (in order):
+/// 1. `type`: The _full_ [type name]
+/// 2. `value`: The serialized value of the reflected type
+///
+/// If the type is already known and the [`TypeInfo`] for it can be retrieved,
+/// [`TypedReflectDeserializer`] may be used instead to avoid requiring these entries.
+///
+/// [`Box<dyn Reflect>`]: crate::Reflect
+/// [`DynamicStruct`]: crate::DynamicStruct
+/// [`DynamicList`]: crate::DynamicList
+/// [`FromReflect`]: crate::FromReflect
+/// [type name]: std::any::type_name
+pub struct UntypedReflectDeserializer<'a> {
+    registry: &'a TypeRegistry,
+}
+
+impl<'a> UntypedReflectDeserializer<'a> {
+    pub fn new(registry: &'a TypeRegistry) -> Self {
+        Self { registry }
     }
+}
 
-    fn visit_f32<E>(self, v: f32) -> Result<Self::Value, E>
+impl<'a, 'de> DeserializeSeed<'de> for UntypedReflectDeserializer<'a> {
+    type Value = Box<dyn Reflect>;
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
     where
-        E: de::Error,
+        D: serde::Deserializer<'de>,
     {
-        Ok(Box::new(v))
+        deserializer.deserialize_any(UntypedReflectDeserializerVisitor {
```

Comment:
Could this be `deserialize_map`? `UntypedReflectDeserializerVisitor` doesn't seem to handle anything except a map.

`deserialize_any` doesn't work with formats that aren't self-describing like ex. bincode.

Reply:
I mentioned this in #4561 (Changelog section), but I'm avoiding this change because I don't think it works even with `deserialize_map`. There might be other problems preventing it, but im not sure what those are. 

For now, I’m leaving it like this so crates like `bincode` can at least recognize it and throw a helpful error message. Making this format binary safe should probably be its own PR.